run_tests.py

```diff
@@ -80,24 +83,47 @@ class GRREverythingTestLoader(test_lib.GRRTestLoader):
   base_class = test_lib.GRRBaseTest
 
 
+class GRROneTestLoader(test_lib.GRRTestLoader):
+  """Load one specific GRR test case."""
+  base_class = test_lib.GRRBaseTest
+
+  def loadTestsFromModule(self, _):
+    """Just return all the tests as if they were in the same module."""
+    test_cases = [
+        self.loadTestsFromTestCase(x) for x in self.base_class.classes.values()
+        if x.__name__ == self.test_class_name]
+
+    return self.suiteClass(test_cases)
+
+  def __init__(self, test_class_name):
+    self.test_class_name = test_class_name
+
+    super(GRROneTestLoader, self).__init__()
+
+
 def RunTest(test_suite, stream=None):
   out_fd = stream
   if stream:
     out_fd = StringIO.StringIO()
 
   try:
-    test_lib.GrrTestProgram(argv=[sys.argv[0], test_suite],
-                            testLoader=GRREverythingTestLoader(
-                                labels=flags.FLAGS.labels),
-                            testRunner=unittest.TextTestRunner(
-                                stream=out_fd))
+    program = test_lib.GrrTestProgram(
+                argv=[sys.argv[0], test_suite],
+                testLoader=GRROneTestLoader(
+                    test_class_name=test_suite),
+                testRunner=unittest.TextTestRunner(
+                    stream=out_fd))
+  except SystemExit as e:
+    # must include this section, as GrrTestProgram issues
```

Comment:
The existing test script can run in two modes - inline: That is really only designed to run one test at the time and its used for debugging mostly (i.e. so you can put break points etc). The single process mode is really designed to run interactively only.

The script is also able to run in multiprocessing mode (which we usually use) and thats used to run the entire suite.

It appears Travis cant run more than one test at the time due to resource issues, so we need to reduce the number of concurrent tests to one - currently running the tests inline is not really a supported mode of operation and the code specifically checks for processes == 1 and goes into that code path. You claim that this is not reliable since some tests are exiting so I am suggesting we stick to the multiprocess mode since it is the current way things work and will require minimal new code.

All the grr test programs are stand alone (i.e. they have their own main()) and are actually designed to run stand alone. This is the best practice for running tests because you ensure that you have test isoltion when you start a new process each time (you can not guarantee this if you dont because we have a lot of shared globals like config library etc). In Rekall we actually spawn the test binary itself but I think in GRR we fork (maybe we should spawn).

Not all tests should run as a separate program but test suites (which are collections of tests) should.

Individual tests are invocable on their own by simply running the test program. They can also be specified as an arg to the run_tests.py script but this is just a convenience. I cant comment on nose since I never used it and I think changing the test framework is probably outside the scope of this cl.

I just suggested to remove that processes == 1 condition from the top and just see if that improves things on travis by running the test program in a child process.

Note that tests can fail also due to a seg fault and catching the exception wont help here.